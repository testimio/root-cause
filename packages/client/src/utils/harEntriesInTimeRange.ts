import type { Har } from 'har-format';

export function harEntriesInTimeRange(harFile: Har, startTime: Date, endTime: Date) {
  return harFile.log.entries.filter((entry) => {
    // native date parsing
    const entryStartDate = new Date(entry.startedDateTime);

    // entry dose not have clear "end" time, so we just go with if the request have started in the current step
    // todo: Try to filter every request that starts/inflight-ends during current step
    return entryStartDate > startTime && entryStartDate < endTime;
  });
}

export function harInTimeRange(harFile: Har, startTime: Date, endTime: Date) {
  const entries = harFile.log.entries.filter((entry) => {
    // native date parsing
    const entryStartDate = new Date(entry.startedDateTime);

    // entry dose not have clear "end" time, so we just go with if the request have started in the current step
    // todo: Try to filter every request that starts/inflight-ends during current step
    return entryStartDate > startTime && entryStartDate < endTime;
  });

  return {
    ...harFile,
    log: {
      ...harFile.log,
      entries: entries,
    },
  };
}
