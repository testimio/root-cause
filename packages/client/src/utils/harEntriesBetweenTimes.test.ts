import { harEntriesInTimeRange } from './harEntriesInTimeRange';
import harContent from './fixtures/forHarEntriesInTimeRange/networkLogs.har.json';
import testResults from './fixtures/forHarEntriesInTimeRange/results.json';
import type { Har, Entry } from 'har-format';

// @ts-ignore
const harContentWithType: Har = harContent;

function entryInterestingParts(entry: Entry) {
  return {
    url: entry.request.url,
    method: entry.request.method,
    startTime: entry.startedDateTime,
    status: entry.response.status,
    statusText: entry.response.statusText,
  };
}

describe('harEntriesInTimeRange', () => {
  test('First step from fixture goto http://jsbin.testim.io/ces', () => {
    const firstStep = testResults.steps[0];

    const { startTimestamp, endTimestamp } = firstStep;

    const startDate = new Date(startTimestamp);
    const endDate = new Date(endTimestamp);
    const r = harEntriesInTimeRange(harContentWithType, startDate, endDate).map(
      entryInterestingParts
    );

    expect(r).toMatchSnapshot();
  });

  test('click [data-job=GET_OK]', () => {
    const firstStep = testResults.steps[1];

    const { startTimestamp, endTimestamp } = firstStep;

    const startDate = new Date(startTimestamp);
    const endDate = new Date(endTimestamp);
    const r = harEntriesInTimeRange(harContentWithType, startDate, endDate).map(
      entryInterestingParts
    );

    expect(r).toMatchSnapshot();
  });

  test('click [data-job=GET_404]', () => {
    const firstStep = testResults.steps[2];

    const { startTimestamp, endTimestamp } = firstStep;

    const startDate = new Date(startTimestamp);
    const endDate = new Date(endTimestamp);
    const r = harEntriesInTimeRange(harContentWithType, startDate, endDate).map(
      entryInterestingParts
    );

    expect(r).toMatchSnapshot();
  });

  test('click [data-job=POST_OK]', () => {
    const firstStep = testResults.steps[3];

    const { startTimestamp, endTimestamp } = firstStep;

    const startDate = new Date(startTimestamp);
    const endDate = new Date(endTimestamp);
    const r = harEntriesInTimeRange(harContentWithType, startDate, endDate).map(
      entryInterestingParts
    );

    expect(r).toMatchSnapshot();
  });

  test('click [data-job=POST_404]', () => {
    const firstStep = testResults.steps[4];

    const { startTimestamp, endTimestamp } = firstStep;

    const startDate = new Date(startTimestamp);
    const endDate = new Date(endTimestamp);
    const r = harEntriesInTimeRange(harContentWithType, startDate, endDate).map(
      entryInterestingParts
    );

    expect(r).toMatchSnapshot();
  });
});
