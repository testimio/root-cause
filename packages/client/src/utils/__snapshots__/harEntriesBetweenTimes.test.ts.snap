// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`harEntriesInTimeRange First step from fixture goto http://jsbin.testim.io/ces 1`] = `
Array [
  Object {
    "method": "GET",
    "startTime": "2020-08-08T13:43:54.128Z",
    "status": 200,
    "statusText": "OK",
    "url": "http://jsbin.testim.io/ces",
  },
  Object {
    "method": "GET",
    "startTime": "2020-08-08T13:43:54.741Z",
    "status": 200,
    "statusText": "OK",
    "url": "http://jsbin-static.testim.io/js/render/edit.js?4.1.4",
  },
  Object {
    "method": "GET",
    "startTime": "2020-08-08T13:43:55.367Z",
    "status": 200,
    "statusText": "OK",
    "url": "http://jsbin-static.testim.io/css/edit.css",
  },
  Object {
    "method": "GET",
    "startTime": "2020-08-08T13:43:55.367Z",
    "status": 200,
    "statusText": "OK",
    "url": "http://jsbin-static.testim.io/images/favicon.png",
  },
]
`;

exports[`harEntriesInTimeRange click [data-job=GET_404] 1`] = `
Array [
  Object {
    "method": "GET",
    "startTime": "2020-08-08T13:43:57.042Z",
    "status": 404,
    "statusText": "",
    "url": "https://postman-echo.com/status/404",
  },
]
`;

exports[`harEntriesInTimeRange click [data-job=GET_OK] 1`] = `
Array [
  Object {
    "method": "GET",
    "startTime": "2020-08-08T13:43:55.962Z",
    "status": 200,
    "statusText": "",
    "url": "https://postman-echo.com/status/200",
  },
]
`;

exports[`harEntriesInTimeRange click [data-job=POST_404] 1`] = `
Array [
  Object {
    "method": "POST",
    "startTime": "2020-08-08T13:43:59.191Z",
    "status": 404,
    "statusText": "",
    "url": "https://postman-echo.com/status/404",
  },
]
`;

exports[`harEntriesInTimeRange click [data-job=POST_OK] 1`] = `
Array [
  Object {
    "method": "POST",
    "startTime": "2020-08-08T13:43:58.112Z",
    "status": 200,
    "statusText": "",
    "url": "https://postman-echo.com/post",
  },
]
`;
